# Creación del dataset conjunto.
En el presente notebook se describe la creación de un dataset único que aúna los datos sobre el nivel educativo alcanzado (nuestras variables objetivos), el gasto en educación, las emisiones en CO2 y la confianza en las vacunas.

Cargamos los módulos necesarios:

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import xml.etree.ElementTree as ET
import requests
import json
from lxml import etree
from bs4 import BeautifulSoup
import math as m
import csv
#import wget

La siguiente función `df_info` nos permite ver los nombres de las columnas de un dataframe, su tipo y cúantos valores únicos toma y cúales son:

In [2]:
# Show unique values for variables. Discrete? Continuous?
# Select variables
def info(df_name):

    # Consider each columns
    columns_level = list(df_name.columns.values)

    for column in columns_level:
        
        # Display column name
        print("Column: " + column)
        
        # Display column type
        print("Type: " + str(df_name[column].dtype))
        
        # Diplay unique values
        temp  = df_name[column].unique()
        print('Unique values ' + str(len(temp)) + ":")
        print(temp)
        
        print('\n')

**Leemos el dataset `adult_education_level.csv`, que contiene los datos sobre el nivel educativo alcanzado:**

In [3]:
# Read .csv file
df_level = pd.read_csv('adult_education_level.csv')

Mostramos su información:

In [4]:
info(df_level)

Column: LOCATION
Type: object
Unique values 48:
['AUS' 'AUT' 'BEL' 'CAN' 'CZE' 'DNK' 'FIN' 'FRA' 'DEU' 'GRC' 'HUN' 'ISL'
 'IRL' 'ITA' 'JPN' 'KOR' 'LUX' 'MEX' 'NLD' 'NZL' 'NOR' 'POL' 'PRT' 'SVK'
 'ESP' 'SWE' 'CHE' 'TUR' 'GBR' 'USA' 'BRA' 'CHL' 'CHN' 'COL' 'EST' 'IDN'
 'ISR' 'LVA' 'RUS' 'SVN' 'ZAF' 'OAVG' 'ARG' 'SAU' 'IND' 'LTU' 'CRI' 'G20']


Column: INDICATOR
Type: object
Unique values 1:
['EDUADULT']


Column: SUBJECT
Type: object
Unique values 7:
['BUPPSRY' 'TRY' 'UPPSRY' 'TRY_MEN' 'TRY_WOMEN' 'UPPSRY_MEN'
 'UPPSRY_WOMEN']


Column: MEASURE
Type: object
Unique values 1:
['PC_25_64']


Column: FREQUENCY
Type: object
Unique values 1:
['A']


Column: TIME
Type: int64
Unique values 40:
[1989 1991 1993 1994 1995 1997 1998 1999 2000 2001 2002 2003 2004 2005
 2006 2007 2008 2009 2010 2011 2012 2013 2014 2015 2016 2017 2018 2019
 2020 1992 1981 1990 1996 1988 1987 1982 1983 1984 1985 1986]


Column: Value
Type: float64
Unique values 6967:
[44.650639 44.127056 47.159046 ... 32.03635  31.42753

In [131]:
df_level.head()

LOCATION INDICATOR  SUBJECT   MEASURE FREQUENCY  TIME      Value  Flag Codes
0      AUS  EDUADULT  BUPPSRY  PC_25_64         A  1989  44.650639         NaN
1      AUS  EDUADULT  BUPPSRY  PC_25_64         A  1991  44.127056         NaN
2      AUS  EDUADULT  BUPPSRY  PC_25_64         A  1993  47.159046         NaN
3      AUS  EDUADULT  BUPPSRY  PC_25_64         A  1994  49.802025         NaN
4      AUS  EDUADULT  BUPPSRY  PC_25_64         A  1995  44.935852         NaN

**Leemos el dataset `education_spending.csv`, que contiene los datos sobre el gasto educativo:**

In [5]:
# Read .csv file
df_spend = pd.read_csv('education_spending.csv')

Mostramos su información:

In [6]:
info(df_spend)

Column: LOCATION
Type: object
Unique values 48:
['AUS' 'AUT' 'BEL' 'CAN' 'CZE' 'DNK' 'FIN' 'FRA' 'DEU' 'GRC' 'HUN' 'ISL'
 'IRL' 'ITA' 'JPN' 'KOR' 'LUX' 'MEX' 'NLD' 'NZL' 'NOR' 'POL' 'PRT' 'SVK'
 'ESP' 'SWE' 'CHE' 'TUR' 'GBR' 'USA' 'BRA' 'CHL' 'CHN' 'COL' 'EST' 'IDN'
 'ISR' 'LVA' 'RUS' 'SVN' 'ZAF' 'OAVG' 'ARG' 'SAU' 'IND' 'LTU' 'CRI' 'G20']


Column: INDICATOR
Type: object
Unique values 1:
['EDUADULT']


Column: SUBJECT
Type: object
Unique values 7:
['BUPPSRY' 'TRY' 'UPPSRY' 'TRY_MEN' 'TRY_WOMEN' 'UPPSRY_MEN'
 'UPPSRY_WOMEN']


Column: MEASURE
Type: object
Unique values 1:
['PC_25_64']


Column: FREQUENCY
Type: object
Unique values 1:
['A']


Column: TIME
Type: int64
Unique values 40:
[1989 1991 1993 1994 1995 1997 1998 1999 2000 2001 2002 2003 2004 2005
 2006 2007 2008 2009 2010 2011 2012 2013 2014 2015 2016 2017 2018 2019
 2020 1992 1981 1990 1996 1988 1987 1982 1983 1984 1985 1986]


Column: Value
Type: float64
Unique values 6967:
[44.650639 44.127056 47.159046 ... 32.03635  31.42753

**Leemos el dataset `co-emissions-per-capita.csv`, que contiene los datos sobre las emisiones de CO2 per cápita:**

In [7]:
# Read .csv file
df_co2 = pd.read_csv('co-emissions-per-capita.csv')

Mostramos su información:

In [8]:
info(df_co2)

Column: Entity
Type: object
Unique values 230:
['Afghanistan' 'Africa' 'Albania' 'Algeria' 'Andorra' 'Angola' 'Anguilla'
 'Antigua and Barbuda' 'Argentina' 'Armenia' 'Aruba' 'Asia'
 'Asia (excl. China & India)' 'Australia' 'Austria' 'Azerbaijan' 'Bahamas'
 'Bahrain' 'Bangladesh' 'Barbados' 'Belarus' 'Belgium' 'Belize' 'Benin'
 'Bermuda' 'Bhutan' 'Bolivia' 'Bonaire Sint Eustatius and Saba'
 'Bosnia and Herzegovina' 'Botswana' 'Brazil' 'British Virgin Islands'
 'Brunei' 'Bulgaria' 'Burkina Faso' 'Burundi' 'Cambodia' 'Cameroon'
 'Canada' 'Cape Verde' 'Central African Republic' 'Chad' 'Chile' 'China'
 'Colombia' 'Comoros' 'Congo' 'Cook Islands' 'Costa Rica' "Cote d'Ivoire"
 'Croatia' 'Cuba' 'Curacao' 'Cyprus' 'Czechia'
 'Democratic Republic of Congo' 'Denmark' 'Djibouti' 'Dominica'
 'Dominican Republic' 'EU-27' 'EU-28' 'Ecuador' 'Egypt' 'El Salvador'
 'Equatorial Guinea' 'Eritrea' 'Estonia' 'Eswatini' 'Ethiopia' 'Europe'
 'Europe (excl. EU-27)' 'Europe (excl. EU-28)' 'Faeroe Islands' 'Fiji

**Leemos el dataset `vaccine.csv`, que contiene los datos sobre las emisiones de CO2 per cápita:**

In [9]:
# Read .csv file
df_vacc = pd.read_csv('vaccine.csv', sep = ";")

Mostramos su información:

In [10]:
info(df_vacc)

Column: country or territory
Type: object
Unique values 149:
['Algeria' 'Benin' 'Botswana' 'Burkina Faso' 'Burundi' 'Cameroon' 'Chad'
 'Comoros' 'Democratic Republic of the Congo' 'Ethiopia' 'Gabon' 'Gambia'
 'Ghana' 'Guinea' 'Ivory Coast' 'Kenya' 'Liberia' 'Madagascar' 'Malawi'
 'Mali' 'Mauritania' 'Mauritius' 'Mozambique' 'Namibia' 'Niger' 'Nigeria'
 'Republic of Congo' 'Rwanda' 'Senegal' 'Sierra Leone' 'South Africa'
 'Swaziland' 'Tanzania' 'Togo' 'Uganda' 'Zambia' 'Zimbabwe' 'Argentina'
 'Bolivia' 'Brazil' 'Canada' 'Chile' 'Colombia' 'Costa Rica'
 'Dominican Republic' 'Ecuador' 'El Salvador' 'Guatemala' 'Haiti'
 'Honduras' 'Mexico' 'Nicaragua' 'Panama' 'Paraguay' 'Peru' 'Uruguay'
 'USA' 'Venezuela' 'Afghanistan' 'Egypt' 'Iran' 'Iraq' 'Jordan' 'Kuwait'
 'Lebanon' 'Libya' 'Morocco' 'Pakistan' 'Palestine' 'Saudi Arabia' 'Syria'
 'Tunisia' 'United Arab Emirates' 'Yemen' 'Albania' 'Armenia' 'Austria'
 'Azerbaijan' 'Belarus' 'Belgium' 'Bosnia and Herzegovina' 'Bulgaria'
 'Croatia' 'Cypru

## En primer lugar, veamos qué países tienen en común todos los dataset:

**Para el dataset relativo al nivel educativo alcanzado (`df_level`)**, observamos que hay 48 países distintos:

In [11]:
unique_location_level = np.unique(df_level['LOCATION'])

In [12]:
len(unique_location_level)

48

In [13]:
unique_location_level

array(['ARG', 'AUS', 'AUT', 'BEL', 'BRA', 'CAN', 'CHE', 'CHL', 'CHN',
       'COL', 'CRI', 'CZE', 'DEU', 'DNK', 'ESP', 'EST', 'FIN', 'FRA',
       'G20', 'GBR', 'GRC', 'HUN', 'IDN', 'IND', 'IRL', 'ISL', 'ISR',
       'ITA', 'JPN', 'KOR', 'LTU', 'LUX', 'LVA', 'MEX', 'NLD', 'NOR',
       'NZL', 'OAVG', 'POL', 'PRT', 'RUS', 'SAU', 'SVK', 'SVN', 'SWE',
       'TUR', 'USA', 'ZAF'], dtype=object)

Observamos que algunos hay dos *países* que llama la atención: 
* OAVG: se corresponde con los datos medios de la OCDE,
* G20: se corresponde con la media de los países del grupo G20.

A priori, estas localizaciónes pueden no estar disponibles en el resto de dataset.

**Para el dataset relativo al gasto educativo `df_spend`**, veamos qué países contiene:

In [14]:
unique_location_spend = np.unique(df_spend['LOCATION'])

In [15]:
len(unique_location_spend)

48

También obtenemos 48 localizaciones (se debe a que la fuente es la misma).

Compobemos que `unique_location_spend` y `unique_location_level` contienen las mismas localizaciones:

In [16]:
# Return locations in unique_location_level that are not in unique_location_spend
[x for x in unique_location_level if x not in unique_location_spend]

[]

In [17]:
# Return locations in unique_location_spend that are not in unique_location_level
[x for x in unique_location_spend if x not in unique_location_level]

[]

Puesto que ambos resultados son listas vacías, concluímos que son el mismpo conjunto. 

**Analizamos los países contenidos en el dataset relativo a las emisiones de CO2 `df_co2`:**

Observamos que hay campos nulos en la columna "Code" y estos se corresponde con continentes o regiones que abarcan a varios países (almacenados en la columna "Entity"):

In [18]:
co2_continents = np.unique(df_co2[df_co2['Code'].isnull() == True]['Entity'])
co2_continents

array(['Africa', 'Asia', 'Asia (excl. China & India)', 'EU-27', 'EU-28',
       'Europe', 'Europe (excl. EU-27)', 'Europe (excl. EU-28)',
       'North America', 'North America (excl. USA)', 'Oceania',
       'South America'], dtype=object)

Comprobemos si alguno de los resultados anteriores está contenido en los países coincidentes entre los de gasto en educación y nivel educativo alcanzado. No los hay:

In [19]:
[x for x in co2_continents if x in unique_location_spend]

[]

Eliminamos las filas que contienen valores nulos en la columna "Code":

In [20]:
# Elkiminamos los que son nulos (ver unicos valores de code al inicio)
unique_location_co2 = np.unique(df_co2[df_co2['Code'].isnull() == False]["Code"])

Miramos qué países de `unique_location_level` no están en `unique_location_co2`:

In [21]:
[x for x in unique_location_level if x not in unique_location_co2]

['G20', 'OAVG']

Tan sólo hay dos no coincidentes. G20, OAVG.

**En el dataset relativo a las vacunas `df_vacc`**, ya observamos que se encuentran los nombres completos de los países y no su código alpha-3. Si bien, en el dataset relativo a las emidiones de CO2 `df_co2`, tenemos dos columnas que nos permitirán solucionar este problema:
* Entity: contiene el nombre completo del pañis;
* Code: contiene el código alpha-3 del país.

Empleando estas dos columnas, concatenamos código alpha-3 y el combre completo del dataset `df_co2`:

In [22]:
co2_code_entity = df_co2['Code'] + '-' + df_co2['Entity']

Lo separamos en tuplas del tipo (codigo alpha-3, nombre completo):

In [23]:
co2_tuples = [tuple(x.split('-')) for x in np.unique([x for x in co2_code_entity if x is not float])]
co2_tuples

[('ABW', 'Aruba'),
 ('AFG', 'Afghanistan'),
 ('AGO', 'Angola'),
 ('AIA', 'Anguilla'),
 ('ALB', 'Albania'),
 ('AND', 'Andorra'),
 ('ARE', 'United Arab Emirates'),
 ('ARG', 'Argentina'),
 ('ARM', 'Armenia'),
 ('ATG', 'Antigua and Barbuda'),
 ('AUS', 'Australia'),
 ('AUT', 'Austria'),
 ('AZE', 'Azerbaijan'),
 ('BDI', 'Burundi'),
 ('BEL', 'Belgium'),
 ('BEN', 'Benin'),
 ('BES', 'Bonaire Sint Eustatius and Saba'),
 ('BFA', 'Burkina Faso'),
 ('BGD', 'Bangladesh'),
 ('BGR', 'Bulgaria'),
 ('BHR', 'Bahrain'),
 ('BHS', 'Bahamas'),
 ('BIH', 'Bosnia and Herzegovina'),
 ('BLR', 'Belarus'),
 ('BLZ', 'Belize'),
 ('BMU', 'Bermuda'),
 ('BOL', 'Bolivia'),
 ('BRA', 'Brazil'),
 ('BRB', 'Barbados'),
 ('BRN', 'Brunei'),
 ('BTN', 'Bhutan'),
 ('BWA', 'Botswana'),
 ('CAF', 'Central African Republic'),
 ('CAN', 'Canada'),
 ('CHE', 'Switzerland'),
 ('CHL', 'Chile'),
 ('CHN', 'China'),
 ('CIV', "Cote d'Ivoire"),
 ('CMR', 'Cameroon'),
 ('COD', 'Democratic Republic of Congo'),
 ('COG', 'Congo'),
 ('COK', 'Cook Isla

Nos quedamos con las localizaciones coincidentes entre las `df_level` (y por tanto `df_spend`) y las que acabamos de obtener:

In [29]:
co2_level_tuples = [x for x in co2_tuples if x[0] in unique_location_level]
co2_level_tuples

[('ARG', 'Argentina'),
 ('AUS', 'Australia'),
 ('AUT', 'Austria'),
 ('BEL', 'Belgium'),
 ('BRA', 'Brazil'),
 ('CAN', 'Canada'),
 ('CHE', 'Switzerland'),
 ('CHL', 'Chile'),
 ('CHN', 'China'),
 ('COL', 'Colombia'),
 ('CRI', 'Costa Rica'),
 ('CZE', 'Czechia'),
 ('DEU', 'Germany'),
 ('DNK', 'Denmark'),
 ('ESP', 'Spain'),
 ('EST', 'Estonia'),
 ('FIN', 'Finland'),
 ('FRA', 'France'),
 ('GBR', 'United Kingdom'),
 ('GRC', 'Greece'),
 ('HUN', 'Hungary'),
 ('IDN', 'Indonesia'),
 ('IND', 'India'),
 ('IRL', 'Ireland'),
 ('ISL', 'Iceland'),
 ('ISR', 'Israel'),
 ('ITA', 'Italy'),
 ('JPN', 'Japan'),
 ('KOR', 'South Korea'),
 ('LTU', 'Lithuania'),
 ('LUX', 'Luxembourg'),
 ('LVA', 'Latvia'),
 ('MEX', 'Mexico'),
 ('NLD', 'Netherlands'),
 ('NOR', 'Norway'),
 ('NZL', 'New Zealand'),
 ('POL', 'Poland'),
 ('PRT', 'Portugal'),
 ('RUS', 'Russia'),
 ('SAU', 'Saudi Arabia'),
 ('SVK', 'Slovakia'),
 ('SVN', 'Slovenia'),
 ('SWE', 'Sweden'),
 ('TUR', 'Turkey'),
 ('USA', 'United States'),
 ('ZAF', 'South Africa')]

In [30]:
len(co2_level_tuples)

46

Seleccionamos las localizaciones entre estos datset y el de las vacunas, que es el único que resta:

In [25]:
vaccines_location = np.unique(df_vacc['country or territory'])

In [26]:
common_locations = [x for x in vaccines_location if x in [x[1] for x in co2_level_tuples]]
common_locations

['Argentina',
 'Australia',
 'Austria',
 'Belgium',
 'Brazil',
 'Canada',
 'Chile',
 'China',
 'Colombia',
 'Costa Rica',
 'Denmark',
 'Estonia',
 'Finland',
 'France',
 'Germany',
 'Greece',
 'Hungary',
 'Iceland',
 'India',
 'Indonesia',
 'Ireland',
 'Israel',
 'Italy',
 'Japan',
 'Latvia',
 'Lithuania',
 'Luxembourg',
 'Mexico',
 'Netherlands',
 'New Zealand',
 'Norway',
 'Poland',
 'Portugal',
 'Russia',
 'Saudi Arabia',
 'Slovakia',
 'Slovenia',
 'South Africa',
 'South Korea',
 'Spain',
 'Sweden',
 'Switzerland',
 'Turkey']

In [27]:
len(common_locations)

43

Países no contenidos:

In [36]:
diff_locations = [x for x in [x[1] for x in co2_level_tuples] if x not in vaccines_location]
diff_locations

['Czechia', 'United Kingdom', 'United States']

Ya tenemos los paises coincidentes entre todos los datsets, que son 43. Pero no están filtrados por año. Esto no nos garantiza que haya datos para todos estos países, en todos los dataset para todos los años disponibles.

## En segundo lugar, filtramos por año:

**Para el datset relativo al nivel educativo `df_level`**, seleccionamos el año 2018 y despreciamos las columnas que no nos interesan.

Recordemos que, en el dataset `df_level`, obteníamos 7 posibles valoes de la columna "SUBJJECT", que mide recoge las variables obtetivo. Por tanto, podemos seleccionar por año, agrupas por país y usar la dfunción `sum()` para comprobar si hay datos repetidos. Si el resultado de "TIME" es 2018x7, es que están todos los datos completos; si es mayor que  2018x7 indica que hay campos duplicados; si es menor que  2018x7 indica que faltan datos.

In [50]:
temp = df_level[df_level['TIME'] == 2018].groupby("LOCATION").sum()['TIME'] 

In [51]:
temp[temp / 2018 < 7]

LOCATION
JPN    6054
Name: TIME, dtype: int64

In [52]:
temp[temp / 2018 > 7]

Series([], Name: TIME, dtype: int64)

Tan sólo nos faltan datos par JPN, algo que comprobaremos a continuación.

Por tanto, como función de agregado en pivot_tables, empleamos "first" ya que no hay países duplicados. Despreciamos ciertas columns que no snos interesan por tomar un único valor que no nos aporta información:

In [53]:
df_1 = df_level[df_level['TIME'] == 2018].\
     drop(['FREQUENCY', 'Flag Codes', 'INDICATOR', 'MEASURE', 'TIME'], axis = 1).\
        pivot_table(values  = 'Value',
                    index   = 'LOCATION',
                    columns = 'SUBJECT',
                    aggfunc = 'first')

Puesto que los nombres de las columas de `df_level` coinciden con los de `df_spend`, añadimos un sufijo:

In [54]:
df_1.columns = df_1.columns + '-level'

In [55]:
df_1

SUBJECT   BUPPSRY-level  TRY-level  TRY_MEN-level  TRY_WOMEN-level  \
LOCATION                                                             
ARG           36.359135  35.663551      31.336042        39.529755   
AUS           18.108641  45.727478      40.441612        50.837093   
AUT           14.702526  32.711426      33.656857        31.769451   
BEL           21.773218  40.638546      36.949390        44.334103   
BRA           46.954311  18.429998      15.601707        20.977964   
CAN            8.377914  57.888363      51.314053        64.394012   
CHE           11.598805  43.744415      47.688629        39.743084   
COL           44.813713  23.391361      21.082104        25.569864   
CRI           60.770302  22.601564      21.556192        23.628597   
CZE            6.139409  24.262077      22.450584        26.133144   
DEU           13.324821  29.065134      31.651323        26.440598   
DNK           18.940752  39.447910      33.794849        45.166317   
ESP           39.884182  37.252804      34.640388        39.846661   
EST           10.810427  41.130722      30.904222        51.221062   
FIN           10.926194  45.188133      37.464184        53.070175   
FRA           20.640024  36.897491      34.210983        39.442608   
G20           34.080354  33.640558      31.928621        35.220310   
GBR           20.707432  45.781677      43.751423        47.731743   
GRC           26.630527  31.739904      30.345835        33.103279   
HUN           15.138174  25.101177      21.622774        28.486479   
IRL           16.825432  46.935135      43.093826        50.639732   
ISL           24.459627  40.279774      32.074547        49.068707   
ISR           13.634555  50.201576      45.497807        54.747131   
ITA           38.252495  19.323732      16.485533        22.103590   
JPN                 NaN  51.928062      51.028938        52.841282   
KOR           11.811645  49.008511      51.760632        46.163258   
LTU            7.022464  41.654594      33.979832        48.837761   
LUX           22.764822  43.886856      43.216248        44.569466   
LVA           12.368318  33.939289      24.911245        42.313160   
MEX           60.853703  17.979799      18.878956        17.188690   
NLD           20.976793  38.343895      37.782265        38.904495   
NOR           17.522373  43.575832      39.068577        48.301937   
NZL           19.542797  39.292675      36.308079        42.103603   
OAVG          22.027774  36.846518      33.577753        40.070351   
POL            7.552199  30.919592      25.205914        36.574348   
PRT           50.189827  24.984337      19.692612        29.829231   
RUS            4.802298  56.725521      49.234489        63.460545   
SVK            8.421120  24.583420      20.983791        28.216055   
SVN           11.862703  32.455208      26.375181        38.859341   
SWE           16.712725  43.247318      36.428112        50.303387   
TUR           59.763054  20.780222      21.766516        19.684227   
USA            9.176164  47.431145      44.312847        50.435326   
ZAF           55.297176  15.387996      14.855694        15.901568   

SUBJECT   UPPSRY-level  UPPSRY_MEN-level  UPPSRY_WOMEN-level  
LOCATION                                                      
ARG          27.977314         28.792295           27.249208  
AUS          36.163879         41.903530           30.615618  
AUT          52.586048         55.005234           50.175697  
BEL          37.588238         40.415058           34.756508  
BRA          34.615692         34.144085           35.040554  
CAN          33.733723         38.883816           28.637413  
CHE          44.656780         41.725693           47.630314  
COL          31.794924         31.868658           31.725365  
CRI          16.628136         16.384611           16.867390  
CZE          69.598511         72.602074           66.496178  
DEU          57.610046         56.017559           59.226143  
DNK          41.611340        

In [56]:
len(df_1)

43

Observamos que hay valores nan para JPN.

Al filtrar datos de nivel educativo por año 2018, ¿estamos perdiendo algún país? Hemos perdido:

In [57]:
[x for x in unique_location_level if x not in df_1.index]

['CHL', 'CHN', 'IDN', 'IND', 'SAU']

In [58]:
len(df_1.index)

43

Hemos perdido 5 países que no tienen datos para 2018.

Repetimos el mismo proceso para el **dataset de gasto educativo `df_spend`**:

Comprobamos si hay duplicados:

In [59]:
temp = df_spend[df_spend['TIME'] == 2018].groupby("LOCATION").sum()['TIME'] 

In [60]:
temp[temp / 2018 < 7]

LOCATION
JPN    6054
Name: TIME, dtype: int64

In [61]:
temp[temp / 2018 > 7]

Series([], Name: TIME, dtype: int64)

Nuevamente, faltan datos para JPN, pero no hay duplicados, luego volvemos a usar la función de agregado "first":

In [62]:
df_2 = df_spend[df_spend['TIME'] == 2018].\
     drop(['FREQUENCY', 'Flag Codes', 'INDICATOR', 'MEASURE', 'TIME'], axis = 1).\
        pivot_table(values = 'Value',
                          index = 'LOCATION',
                          columns='SUBJECT',
                          aggfunc='first')

Añadimos un sufijo a los nombres de las columnas:

In [63]:
df_2.columns = df_2.columns + '-spend'

In [64]:
df_2

SUBJECT   BUPPSRY-spend  TRY-spend  TRY_MEN-spend  TRY_WOMEN-spend  \
LOCATION                                                             
ARG           36.359135  35.663551      31.336042        39.529755   
AUS           18.108641  45.727478      40.441612        50.837093   
AUT           14.702526  32.711426      33.656857        31.769451   
BEL           21.773218  40.638546      36.949390        44.334103   
BRA           46.954311  18.429998      15.601707        20.977964   
CAN            8.377914  57.888363      51.314053        64.394012   
CHE           11.598805  43.744415      47.688629        39.743084   
COL           44.813713  23.391361      21.082104        25.569864   
CRI           60.770302  22.601564      21.556192        23.628597   
CZE            6.139409  24.262077      22.450584        26.133144   
DEU           13.324821  29.065134      31.651323        26.440598   
DNK           18.940752  39.447910      33.794849        45.166317   
ESP           39.884182  37.252804      34.640388        39.846661   
EST           10.810427  41.130722      30.904222        51.221062   
FIN           10.926194  45.188133      37.464184        53.070175   
FRA           20.640024  36.897491      34.210983        39.442608   
G20           34.080354  33.640558      31.928621        35.220310   
GBR           20.707432  45.781677      43.751423        47.731743   
GRC           26.630527  31.739904      30.345835        33.103279   
HUN           15.138174  25.101177      21.622774        28.486479   
IRL           16.825432  46.935135      43.093826        50.639732   
ISL           24.459627  40.279774      32.074547        49.068707   
ISR           13.634555  50.201576      45.497807        54.747131   
ITA           38.252495  19.323732      16.485533        22.103590   
JPN                 NaN  51.928062      51.028938        52.841282   
KOR           11.811645  49.008511      51.760632        46.163258   
LTU            7.022464  41.654594      33.979832        48.837761   
LUX           22.764822  43.886856      43.216248        44.569466   
LVA           12.368318  33.939289      24.911245        42.313160   
MEX           60.853703  17.979799      18.878956        17.188690   
NLD           20.976793  38.343895      37.782265        38.904495   
NOR           17.522373  43.575832      39.068577        48.301937   
NZL           19.542797  39.292675      36.308079        42.103603   
OAVG          22.027774  36.846518      33.577753        40.070351   
POL            7.552199  30.919592      25.205914        36.574348   
PRT           50.189827  24.984337      19.692612        29.829231   
RUS            4.802298  56.725521      49.234489        63.460545   
SVK            8.421120  24.583420      20.983791        28.216055   
SVN           11.862703  32.455208      26.375181        38.859341   
SWE           16.712725  43.247318      36.428112        50.303387   
TUR           59.763054  20.780222      21.766516        19.684227   
USA            9.176164  47.431145      44.312847        50.435326   
ZAF           55.297176  15.387996      14.855694        15.901568   

SUBJECT   UPPSRY-spend  UPPSRY_MEN-spend  UPPSRY_WOMEN-spend  
LOCATION                                                      
ARG          27.977314         28.792295           27.249208  
AUS          36.163879         41.903530           30.615618  
AUT          52.586048         55.005234           50.175697  
BEL          37.588238         40.415058           34.756508  
BRA          34.615692         34.144085           35.040554  
CAN          33.733723         38.883816           28.637413  
CHE          44.656780         41.725693           47.630314  
COL          31.794924         31.868658           31.725365  
CRI          16.628136         16.384611           16.867390  
CZE          69.598511         72.602074           66.496178  
DEU          57.610046         56.017559           59.226143  
DNK          41.611340        

In [65]:
len(df_2)

43

AL filtrar datos de gasto educativo por año 2018, ¿estamos perdiendo algún país? Hemos perdido:

In [66]:
[x for x in unique_location_spend if x not in df_2.index]

['CHL', 'CHN', 'IDN', 'IND', 'SAU']

In [67]:
len(df_1.index)

43

Hemos perdido los mismos países que en el dataset de nivel eductivo.

**Filtramos por año 2018 en el dataset relativo al C02:**

In [68]:
#[x for x in co2_level_tuples if x[1] in vaccines_location]

Originalmente, sin filtrar por año, coincidían 46 países entre los dataset de la OCDE y el de CO2 de un total de 48 países (todos salvo G20 y OAVG). Filtrando por 2018 en los dataset de la OCDE, tenemos 43 paises. Como ninguno de los paises que hemos perdido es G20 U OAVG, entonces, como mucho, habrá 41 paises en CO2 para el año 2018.

In [69]:
definite_countries = [x[0] for x in co2_level_tuples if x[1] in vaccines_location]

In [70]:
df_co2_countries = df_co2[df_co2['Code'].isin(definite_countries)]

In [71]:
df_3 = df_co2_countries[df_co2_countries['Year'] == 2018].\
                    drop(['Entity', 'Year'], axis = 1).\
                    set_index('Code')

In [72]:
df_3

Annual CO2 emissions (per capita)
Code                                   
ARG                              4.0824
AUS                             16.7081
AUT                              7.4865
BEL                              8.7289
BRA                              2.3091
CAN                             15.6299
CHL                              4.5068
CHN                              7.2077
COL                              1.7092
CRI                              1.6446
DNK                              6.0277
EST                             13.5644
FIN                              8.3021
FRA                              4.9603
DEU                              9.0721
GRC                              6.8244
HUN                              5.0897
ISL                             10.8799
IND                              1.9220
IDN                              2.2972
IRL                              8.1340
ISR                              7.2238
ITA                              5.7568
JPN                              8.9893
LVA                              4.0755
LTU                              4.8893
LUX                             15.8354
MEX                              3.6081
NLD                              9.3215
NZL                              7.5455
NOR                              8.3307
POL                              8.8866
PRT                              5.0206
RUS                             11.6078
SAU                             18.4541
SVK                              6.6073
SVN                              6.9753
ZAF                              8.0321
KOR                             13.1250
ESP                              5.7763
SWE                              4.2215
CHE                              4.3243
TUR                              5.0940

Por comodidad, nemobramos la columnas:

In [73]:
df_3.columns = ['C02']

Por comodidad, renombramos el índice:

In [74]:
df_3.index.name = "LOCATION"

Obtenemos:

In [75]:
len(df_3.index)

43

Si bien, aún no hemos filtrado por país:

In [76]:
[x for x in df_1.index if x not in df_3.index]

['CZE', 'G20', 'GBR', 'OAVG', 'USA']

Los países de la celda anterior no son coincidentes.

In [77]:
# 43 - 5 = 38 paises

No sólo hemos perdido G20 y OAVG sino también CZE, GBR, USA.

In [78]:
len([x for x in df_1.index if x in df_3.index])

38

In [79]:
my_list_countries_ocde_co2 = [x for x in df_1.index if x in df_3.index]

**Filtramos por año 2018 en vacunas:**

Empleamos el mapeto que hicimos con las tuplas (codigo alpha-3, nombre completo) de los países:

In [80]:
definite_countries = [x for x in co2_level_tuples if x[1] in vaccines_location]

In [81]:
definite_countries_name = [x[1] for x in co2_level_tuples if x[1] in vaccines_location]
len(definite_countries_name)

43

In [82]:
df_vacc_filt = df_vacc[df_vacc['country or territory'].isin(definite_countries_name)]

In [83]:
for x in definite_countries:
    for y in df_vacc_filt['country or territory'].keys():
        if x[1] == df_vacc_filt['country or territory'][y]:
             df_vacc_filt['country or territory'][y] = x[0]

<ipython-input-83-eefc1d3fda74>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_vacc_filt['country or territory'][y] = x[0]
C:\Users\usuario\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3437: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  exec(code_obj, self.user_global_ns, self.user_ns)


In [84]:
df_vacc_filt

country or territory who_region  count  sagree  sdisagree   question  \
123                  ZAF        AFR   1157     806         30  important   
124                  ZAF        AFR   1157     562         48       safe   
125                  ZAF        AFR   1157     588         49  effective   
126                  ZAF        AFR    421     346         14  important   
127                  ZAF        AFR    421     287         16       safe   
..                   ...        ...    ...     ...        ...        ...   
853                  KOR        WPR    898     143          9       safe   
854                  KOR        WPR    898     211          7  effective   
855                  KOR        WPR   1500     604          8  important   
856                  KOR        WPR   1500     486         15       safe   
857                  KOR        WPR   1500     522          7  effective   

            time  
123  2015,833333  
124  2015,833333  
125  2015,833333  
126  2018,513889  
127  2018,513889  
..           ...  
853  2018,580556  
854  2018,580556  
855  2019,947945  
856  2019,947945  
857  2019,947945  

[336 rows x 7 columns]

La columna "Time" está guardada como string. Puesto que sólo nos interesa el año y no el año y el mes llevaremos a cabo el siguiente proceso:
* Sustituimos la coma (,) por un punto (.)
* Convertimos a float;
* Redondeamos hacia abajo.

In [85]:
#list(map(lambda x: x.replace(',', '.'), df_vacc_filt['time']))

In [86]:
#list(map(lambda x: m.floor(float(x.replace(',', '.'))), df_vacc_filt['time']))

In [87]:
df_vacc_filt['time'] = list(map(lambda x: m.floor(float(x.replace(',', '.'))), df_vacc_filt['time']))

<ipython-input-87-f04e1ff0dc36>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_vacc_filt['time'] = list(map(lambda x: m.floor(float(x.replace(',', '.'))), df_vacc_filt['time']))


Salta un earning porque estamos modificando en el original, pero no nos supone ningún perjuicio. De este modo, la columna "time" resulta:

In [88]:
df_vacc_filt.head()

country or territory who_region  count  sagree  sdisagree   question  time
123                  ZAF        AFR   1157     806         30  important  2015
124                  ZAF        AFR   1157     562         48       safe  2015
125                  ZAF        AFR   1157     588         49  effective  2015
126                  ZAF        AFR    421     346         14  important  2018
127                  ZAF        AFR    421     287         16       safe  2018

Filtramos por el año 2018 y agrupamos según territorio. Para ver queé función de agregado utilizar, repetimos el proceso ya explicado para ver si hay países que realizaron más de una encuesta en 2018. Esta vez, tenemos 3 valores que nos interesan (imprtant, safe, effective) en la columna "question":

In [89]:
temp = df_vacc_filt[df_vacc_filt["time"] == 2018].groupby("country or territory").sum()['time']

Países en los que se hizo una única encuesta en 2018:

In [90]:
temp[temp / 2018 == 3 ]

country or territory
ARG    6054
AUS    6054
BRA    6054
CAN    6054
CHE    6054
CHL    6054
CHN    6054
COL    6054
CRI    6054
IDN    6054
IND    6054
ISL    6054
ISR    6054
JPN    6054
KOR    6054
MEX    6054
NOR    6054
NZL    6054
RUS    6054
SAU    6054
TUR    6054
ZAF    6054
Name: time, dtype: int64

Países en los que se hizo más de una encuesta en 2018:

In [91]:
temp[temp / 2018 > 3]

country or territory
AUT    12108
BEL    12108
DEU    12108
DNK    12108
ESP    12108
EST    12108
FIN    12108
FRA    12108
GRC    12108
HUN    12108
IRL    12108
ITA    12108
LTU    12108
LUX    12108
LVA    12108
NLD    12108
POL    12108
PRT    12108
SVK    12108
SVN    12108
SWE    12108
Name: time, dtype: int64

Países en los que se hizo menos de una encuenta en 2018 (esta comprobaci´pn carece de sentido pero, por si acaso):

In [92]:
temp[temp / 2018 < 3]

Series([], Name: time, dtype: int64)

Por tanto, si empleasemos pivot_tables con función de agregado "first", estaríamos perdiendo información.

Filtramos por año 2018 y despreciamos las columnas que no nos interesa:

In [93]:
df_vacc_2018 = df_vacc_filt[df_vacc_filt['time'] == 2018].\
                drop(['time', 'who_region'], axis = 1)

In [94]:
df_vacc_2018

country or territory  count  sagree  sdisagree   question
126                  ZAF    421     346         14  important
127                  ZAF    421     287         16       safe
128                  ZAF    421     279         24  effective
150                  ARG    962     902          6  important
151                  ARG    962     737         13       safe
..                   ...    ...     ...        ...        ...
823                  NZL    966     399         32       safe
824                  NZL    966     524         14  effective
852                  KOR    898     345         12  important
853                  KOR    898     143          9       safe
854                  KOR    898     211          7  effective

[192 rows x 5 columns]

Empleamos `np.mean` como función de agregado para los así tratas los países en los que se realizó más de una encuenta en 2018:

In [95]:
df_temp = df_vacc_2018.pivot_table(values = ['count', 'sagree', 'sdisagree'],
                          index   = 'country or territory',
                          columns = 'question',
                          aggfunc = np.mean)

In [96]:
df_temp

count                      sagree                    \
question             effective important    safe effective important    safe   
country or territory                                                           
ARG                      962.0     962.0   962.0     702.0     902.0   737.0   
AUS                      988.0     988.0   988.0     707.0     824.0   602.0   
AUT                      942.0     942.0   942.0     423.0     552.0   342.0   
BEL                      970.0     970.0   970.0     486.5     648.5   382.0   
BRA                      950.0     950.0   950.0     546.0     798.0   584.0   
CAN                      985.0     985.0   985.0     569.0     745.0   523.0   
CHE                      930.0     930.0   930.0     298.0     452.0   211.0   
CHL                      914.0     914.0   914.0     577.0     723.0   555.0   
CHN                     3233.0    3233.0  3233.0    1570.0    2093.0  1508.0   
COL                      920.0     920.0   920.0     575.0     810.0   598.0   
CRI                      949.0     949.0   949.0     688.0     893.0   696.0   
DEU                      950.0     950.0   950.0     496.0     641.5   418.5   
DNK                      989.0     989.0   989.0     610.0     801.0   553.0   
ESP                      993.0     993.0   993.0     645.0     779.0   616.0   
EST                      972.5     972.5   972.5     371.5     598.5   299.0   
FIN                      981.5     981.5   981.5     528.0     812.5   465.0   
FRA                      986.0     986.0   986.0     282.0     444.5   211.5   
GRC                      979.5     979.5   979.5     517.5     706.5   451.0   
HUN                      980.5     980.5   980.5     615.5     786.5   607.0   
IDN                      639.0     639.0   639.0     313.0     470.0   390.0   
IND                     2707.0    2707.0  2707.0    2321.0    2405.0  2339.0   
IRL                      978.5     978.5   978.5     528.5     662.0   458.5   
ISL                      485.0     485.0   485.0     358.0     440.0   131.0   
ISR                      998.0     998.0   998.0     487.0     659.0   402.0   
ITA                      992.0     992.0   992.0     508.5     639.5   426.5   
JPN                      939.0     939.0   939.0     134.0     220.0    81.0   
KOR                      898.0     898.0   898.0     211.0     345.0   143.0   
LTU                      891.5     891.5   891.5     295.5     461.5   300.5   
LUX                      730.0     730.0   730.0     318.0     502.0   251.0   
LVA                      998.5     998.5   998.5     248.5     470.0   245.0   
MEX                      949.0     949.0   949.0     721.0     844.0   715.0   
NLD                      949.5     949.5   949.5     445.0     648.5   402.5   
NOR                      898.0     898.0   898.0     636.0     788.0   515.0   
NZL                      966.0     966.0   966.0     524.0     709.0   399.0   
POL                      944.0     944.0   944.0     500.5     556.5   447.0   
PRT                      989.0     989.0   989.0     477.0     759.5   531.0   
RUS                     1828.0    1828.0  1828.0     741.0    1118.0   554.0   
SAU                      963.0     963.0   963.0     655.0     813.0   674.0   
SVK                      966.5     966.5   966.5     451.5     581.0   382.5   
SVN                      980.0     980.0   980.0     394.0     547.0   331.0   
SWE                     1006.5    1006.5  1006.5     509.5     720.5   397.0   
TUR                      895.0     895.0   895.0     599.0     718.0   543.0   
ZAF                      421.0     421.0   421.0     279.0     346.0   287.0   

                     sdisagree                   
question             effective important   safe  
country or territory                             
ARG                        8.0       6.0   13.0  
AUS                       17.0      18.0   33.0  
AUT                       30.0      51.0   55.5  
BEL                       30.0      22.0   6

Comprobemos que ha salido bien, analizando algunos valores. En ZAF, sólo se hizo una encuesta:

In [97]:
df_vacc_2018[df_vacc_2018["country or territory"] == "ZAF"]

country or territory  count  sagree  sdisagree   question
126                  ZAF    421     346         14  important
127                  ZAF    421     287         16       safe
128                  ZAF    421     279         24  effective

In [98]:
df_temp.loc["ZAF"]

           question 
count      effective    421.0
           important    421.0
           safe         421.0
sagree     effective    279.0
           important    346.0
           safe         287.0
sdisagree  effective     24.0
           important     14.0
           safe          16.0
Name: ZAF, dtype: float64

Los datos coinciden.

En España, sí se hizo más de una encuesta:

In [99]:
df_vacc_2018[df_vacc_2018["country or territory"] == "ESP"]

country or territory  count  sagree  sdisagree   question
678                  ESP   1005     823          2  important
679                  ESP   1005     642          6       safe
680                  ESP   1005     703          8  effective
681                  ESP    981     735         22  important
682                  ESP    981     590         24       safe
683                  ESP    981     587         31  effective

In [100]:
df_temp.loc["ESP"]

           question 
count      effective    993.0
           important    993.0
           safe         993.0
sagree     effective    645.0
           important    779.0
           safe         616.0
sdisagree  effective     19.5
           important     12.0
           safe          15.0
Name: ESP, dtype: float64

Comprobamos que la media de encuentados es (1005 + 981) / 2, que está recogida como 993 en el dataset `df_temp`. Del mismo modo, podemos fijarnos que, en en cada una de las encuestas, a la cuestiñon "important", en "sdisagrree", responden 2 y 22 personas, que da una media de 12 y así es como se recoge en `df_temp`.

Lo hemos llamado `df_temp` ya que es temporal, pues tenemos que hacer frente a otro problema: no en todos los pañises se ha preguntado al mismo número de personas. Por ello, tenemos que considerar los datos de frecuencia relativa y no los datos de frecuencia absoluta.

Antes de nada, aplanamos las columasn jerárquitas de `df_temp` y cambiamos el nombre del índice:

In [101]:
[' '.join(col).strip() for col in df_temp.columns.values]

['count effective',
 'count important',
 'count safe',
 'sagree effective',
 'sagree important',
 'sagree safe',
 'sdisagree effective',
 'sdisagree important',
 'sdisagree safe']

In [102]:
df_temp = df_temp.set_axis(['-'.join(col) for col in df_temp.columns.values], axis = 1)

In [103]:
df_temp.index.name = "LOCATION"

In [104]:
df_temp

count-effective  count-important  count-safe  sagree-effective  \
LOCATION                                                                   
ARG                 962.0            962.0       962.0             702.0   
AUS                 988.0            988.0       988.0             707.0   
AUT                 942.0            942.0       942.0             423.0   
BEL                 970.0            970.0       970.0             486.5   
BRA                 950.0            950.0       950.0             546.0   
CAN                 985.0            985.0       985.0             569.0   
CHE                 930.0            930.0       930.0             298.0   
CHL                 914.0            914.0       914.0             577.0   
CHN                3233.0           3233.0      3233.0            1570.0   
COL                 920.0            920.0       920.0             575.0   
CRI                 949.0            949.0       949.0             688.0   
DEU                 950.0            950.0       950.0             496.0   
DNK                 989.0            989.0       989.0             610.0   
ESP                 993.0            993.0       993.0             645.0   
EST                 972.5            972.5       972.5             371.5   
FIN                 981.5            981.5       981.5             528.0   
FRA                 986.0            986.0       986.0             282.0   
GRC                 979.5            979.5       979.5             517.5   
HUN                 980.5            980.5       980.5             615.5   
IDN                 639.0            639.0       639.0             313.0   
IND                2707.0           2707.0      2707.0            2321.0   
IRL                 978.5            978.5       978.5             528.5   
ISL                 485.0            485.0       485.0             358.0   
ISR                 998.0            998.0       998.0             487.0   
ITA                 992.0            992.0       992.0             508.5   
JPN                 939.0            939.0       939.0             134.0   
KOR                 898.0            898.0       898.0             211.0   
LTU                 891.5            891.5       891.5             295.5   
LUX                 730.0            730.0       730.0             318.0   
LVA                 998.5            998.5       998.5             248.5   
MEX                 949.0            949.0       949.0             721.0   
NLD                 949.5            949.5       949.5             445.0   
NOR                 898.0            898.0       898.0             636.0   
NZL                 966.0            966.0       966.0             524.0   
POL                 944.0            944.0       944.0             500.5   
PRT                 989.0            989.0       989.0             477.0   
RUS                1828.0           1828.0      1828.0             741.0   
SAU                 963.0            963.0       963.0             655.0   
SVK                 966.5            966.5       966.5             451.5   
SVN                 980.0            980.0       980.0             394.0   
SWE                1006.5           1006.5      1006.5             509.5   
TUR                 895.0            895.0       895.0             599.0   
ZAF                 421.0            421.0       421.0             279.0   

          sagree-important  sagree-safe  sdisagree-effective  \
LOCATION                                                       
ARG                  902.0        737.0                  8.0   
AUS                  824.0        602.0                 17.0   
AUT                  552.0        342.0                 30.0   
BEL                  648.5        382.0                 30.0   
BRA                  798.0        584.0                 12.0   
CAN                  745.0        523.0                 15.0   
CHE                  452.0        211.0                 11.0   
CHL          

Lo establemcemos como definitivo:

In [105]:
df_4 = df_temp

In [106]:
# PASAMOS A PORCENTAJES

Puesto que las columnas de nombre "count-" se corresponden a la variable original "count·, a priori deben tener valores repetidos para cada fila. Lo comprobamos:

In [107]:
# Comprobamos que todos los contadores son los mismos
cols = df_4.columns.values[:3]
df_4['Same-counter?'] = df_4[cols].eq(df_4[cols[0]], axis=0).all(axis = 1)

In [108]:
np.any(df_4['Same-counter?'] == False)

False

Como todos los counter son iguales, nos quedamos con uno:

In [109]:
df_4 = df_4.drop(['Same-counter?', 'count-effective', 'count-important'], axis = 1)

In [110]:
df_4 = df_4.rename(columns = {'count-safe' : 'count'})

In [111]:
df_4

count  sagree-effective  sagree-important  sagree-safe  \
LOCATION                                                            
ARG        962.0             702.0             902.0        737.0   
AUS        988.0             707.0             824.0        602.0   
AUT        942.0             423.0             552.0        342.0   
BEL        970.0             486.5             648.5        382.0   
BRA        950.0             546.0             798.0        584.0   
CAN        985.0             569.0             745.0        523.0   
CHE        930.0             298.0             452.0        211.0   
CHL        914.0             577.0             723.0        555.0   
CHN       3233.0            1570.0            2093.0       1508.0   
COL        920.0             575.0             810.0        598.0   
CRI        949.0             688.0             893.0        696.0   
DEU        950.0             496.0             641.5        418.5   
DNK        989.0             610.0             801.0        553.0   
ESP        993.0             645.0             779.0        616.0   
EST        972.5             371.5             598.5        299.0   
FIN        981.5             528.0             812.5        465.0   
FRA        986.0             282.0             444.5        211.5   
GRC        979.5             517.5             706.5        451.0   
HUN        980.5             615.5             786.5        607.0   
IDN        639.0             313.0             470.0        390.0   
IND       2707.0            2321.0            2405.0       2339.0   
IRL        978.5             528.5             662.0        458.5   
ISL        485.0             358.0             440.0        131.0   
ISR        998.0             487.0             659.0        402.0   
ITA        992.0             508.5             639.5        426.5   
JPN        939.0             134.0             220.0         81.0   
KOR        898.0             211.0             345.0        143.0   
LTU        891.5             295.5             461.5        300.5   
LUX        730.0             318.0             502.0        251.0   
LVA        998.5             248.5             470.0        245.0   
MEX        949.0             721.0             844.0        715.0   
NLD        949.5             445.0             648.5        402.5   
NOR        898.0             636.0             788.0        515.0   
NZL        966.0             524.0             709.0        399.0   
POL        944.0             500.5             556.5        447.0   
PRT        989.0             477.0             759.5        531.0   
RUS       1828.0             741.0            1118.0        554.0   
SAU        963.0             655.0             813.0        674.0   
SVK        966.5             451.5             581.0        382.5   
SVN        980.0             394.0             547.0        331.0   
SWE       1006.5             509.5             720.5        397.0   
TUR        895.0             599.0             718.0        543.0   
ZAF        421.0             279.0             346.0        287.0   

          sdisagree-effective  sdisagree-important  sdisagree-safe  
LOCATION                                                            
ARG                       8.0                  6.0            13.0  
AUS                      17.0                 18.0            33.0  
AUT                      30.0                 51.0            55.5  
BEL                      30.0                 22.0            64.0  
BRA                      12.0                  2.0            14.0  
CAN                      15.0                 20.0            28.0  
CHE                      11.0                 35.0            36.0  
CHL                      18.0                 16.0            37.0  
CHN                      31.0                 24.0            73.0  
COL                      17.0                  2.0            28.0  
CRI                      22.0                  5.0            18.0  
DEU     

Establecemos los porcentajes:

In [112]:
for name in df_4.columns[1:]:
    df_4[name] = df_4[name] / df_4['count']

In [113]:
df_4

count  sagree-effective  sagree-important  sagree-safe  \
LOCATION                                                            
ARG        962.0          0.729730          0.937630     0.766112   
AUS        988.0          0.715587          0.834008     0.609312   
AUT        942.0          0.449045          0.585987     0.363057   
BEL        970.0          0.501546          0.668557     0.393814   
BRA        950.0          0.574737          0.840000     0.614737   
CAN        985.0          0.577665          0.756345     0.530964   
CHE        930.0          0.320430          0.486022     0.226882   
CHL        914.0          0.631291          0.791028     0.607221   
CHN       3233.0          0.485617          0.647386     0.466440   
COL        920.0          0.625000          0.880435     0.650000   
CRI        949.0          0.724974          0.940991     0.733404   
DEU        950.0          0.522105          0.675263     0.440526   
DNK        989.0          0.616785          0.809909     0.559151   
ESP        993.0          0.649547          0.784491     0.620342   
EST        972.5          0.382005          0.615424     0.307455   
FIN        981.5          0.537952          0.827815     0.473765   
FRA        986.0          0.286004          0.450811     0.214503   
GRC        979.5          0.528331          0.721286     0.460439   
HUN        980.5          0.627741          0.802142     0.619072   
IDN        639.0          0.489828          0.735524     0.610329   
IND       2707.0          0.857407          0.888437     0.864056   
IRL        978.5          0.540112          0.676546     0.468574   
ISL        485.0          0.738144          0.907216     0.270103   
ISR        998.0          0.487976          0.660321     0.402806   
ITA        992.0          0.512601          0.644657     0.429940   
JPN        939.0          0.142705          0.234292     0.086262   
KOR        898.0          0.234967          0.384187     0.159243   
LTU        891.5          0.331464          0.517667     0.337072   
LUX        730.0          0.435616          0.687671     0.343836   
LVA        998.5          0.248873          0.470706     0.245368   
MEX        949.0          0.759747          0.889357     0.753425   
NLD        949.5          0.468668          0.682991     0.423907   
NOR        898.0          0.708241          0.877506     0.573497   
NZL        966.0          0.542443          0.733954     0.413043   
POL        944.0          0.530191          0.589513     0.473517   
PRT        989.0          0.482305          0.767947     0.536906   
RUS       1828.0          0.405361          0.611597     0.303063   
SAU        963.0          0.680166          0.844237     0.699896   
SVK        966.5          0.467150          0.601138     0.395758   
SVN        980.0          0.402041          0.558163     0.337755   
SWE       1006.5          0.506210          0.715847     0.394436   
TUR        895.0          0.669274          0.802235     0.606704   
ZAF        421.0          0.662708          0.821853     0.681710   

          sdisagree-effective  sdisagree-important  sdisagree-safe  
LOCATION                                                            
ARG                  0.008316             0.006237        0.013514  
AUS                  0.017206             0.018219        0.033401  
AUT                  0.031847             0.054140        0.058917  
BEL                  0.030928             0.022680        0.065979  
BRA                  0.012632             0.002105        0.014737  
CAN                  0.015228             0.020305        0.028426  
CHE                  0.011828             0.037634        0.038710  
CHL                  0.019694             0.017505        0.040481  
CHN                  0.009589             0.007423        0.022580  
COL                  0.018478             0.002174        0.030435  
CRI                  0.023182             0.005269        0.018967  
DEU     

Podemos observar que no todo el mundo responde a las preguntas. Hay gente que o no sabe, o no contesta. Por ejemplo, para "ZAF", seleccionamos sagree-effective y sdisagree-effective que deberían sumar 1, pero no es así:

In [114]:
df_4.loc["ZAF"]

count                  421.000000
sagree-effective         0.662708
sagree-important         0.821853
sagree-safe              0.681710
sdisagree-effective      0.057007
sdisagree-important      0.033254
sdisagree-safe           0.038005
Name: ZAF, dtype: float64

In [115]:
sum(df_4.loc["ZAF"][[1,4]])

0.7197149643705463

In [116]:
sum(df_4.loc["ZAF"][[2,5]])

0.8551068883610451

¿Cuántos países hemos perdido?

In [117]:
len(np.unique(df_4.index))

43

In [118]:
[x for x in df_1.index if x not in df_3.index]

['CZE', 'G20', 'GBR', 'OAVG', 'USA']

In [119]:
# 43 - 5 = 38 ! correcto. 

## Por último, unimos todo en un mismo dataset:

Mergeamos `df_1` (nivel educativo) y `df_2` (gasto educativo):

In [120]:
merge_1 = df_1.merge(df_2, left_index = True, right_index = True)

Mergeamos con `df_3` (emisiones CO2):

In [121]:
merge_2 = merge_1.merge(df_3, left_index = True, right_index = True)

In [122]:
len(merge_2)

38

Mergeamos con `df_4` (vacunas):

In [123]:
merge_3 = merge_2.merge(df_4, left_index = True, right_index = True)
len(merge_3)

38

Comprobamos JPN:

In [124]:
merge_3.loc['JPN']

BUPPSRY-level                 NaN
TRY-level               51.928062
TRY_MEN-level           51.028938
TRY_WOMEN-level         52.841282
UPPSRY-level                  NaN
UPPSRY_MEN-level              NaN
UPPSRY_WOMEN-level            NaN
BUPPSRY-spend                 NaN
TRY-spend               51.928062
TRY_MEN-spend           51.028938
TRY_WOMEN-spend         52.841282
UPPSRY-spend                  NaN
UPPSRY_MEN-spend              NaN
UPPSRY_WOMEN-spend            NaN
C02                      8.989300
count                  939.000000
sagree-effective         0.142705
sagree-important         0.234292
sagree-safe              0.086262
sdisagree-effective      0.008520
sdisagree-important      0.011715
sdisagree-safe           0.012780
Name: JPN, dtype: float64

Otra opción, es rellenar los valores vacíos con 0 en lugar de nan:

In [125]:
merge_3_0s = merge_2.join(df_4).fillna(0)

In [126]:
merge_3_0s.loc['JPN']

BUPPSRY-level            0.000000
TRY-level               51.928062
TRY_MEN-level           51.028938
TRY_WOMEN-level         52.841282
UPPSRY-level             0.000000
UPPSRY_MEN-level         0.000000
UPPSRY_WOMEN-level       0.000000
BUPPSRY-spend            0.000000
TRY-spend               51.928062
TRY_MEN-spend           51.028938
TRY_WOMEN-spend         52.841282
UPPSRY-spend             0.000000
UPPSRY_MEN-spend         0.000000
UPPSRY_WOMEN-spend       0.000000
C02                      8.989300
count                  939.000000
sagree-effective         0.142705
sagree-important         0.234292
sagree-safe              0.086262
sdisagree-effective      0.008520
sdisagree-important      0.011715
sdisagree-safe           0.012780
Name: JPN, dtype: float64

### Escribimos los archivos en documentos .csv:

Guardamos en csv:
* target_education_nan.csv: contiene nans; guardamos con índices = código país
* target_education_0.csv: contiene 0s guardamos con índices = código país

In [127]:
merge_3.to_csv('target_education_nan.csv')

In [128]:
merge_3_0s.to_csv('target_education_0.csv')

Porbamos a leer:

In [140]:
final_dset = pd.read_csv('target_education_0.csv')

In [141]:
final_dset.head()

LOCATION  BUPPSRY-level  TRY-level  TRY_MEN-level  TRY_WOMEN-level  \
0      ARG      36.359135  35.663551      31.336042        39.529755   
1      AUS      18.108641  45.727478      40.441612        50.837093   
2      AUT      14.702526  32.711426      33.656857        31.769451   
3      BEL      21.773218  40.638546      36.949390        44.334103   
4      BRA      46.954311  18.429998      15.601707        20.977964   

   UPPSRY-level  UPPSRY_MEN-level  UPPSRY_WOMEN-level  BUPPSRY-spend  \
0     27.977314         28.792295           27.249208      36.359135   
1     36.163879         41.903530           30.615618      18.108641   
2     52.586048         55.005234           50.175697      14.702526   
3     37.588238         40.415058           34.756508      21.773218   
4     34.615692         34.144085           35.040554      46.954311   

   TRY-spend  ...  UPPSRY_MEN-spend  UPPSRY_WOMEN-spend      C02  count  \
0  35.663551  ...         28.792295           27.249208   4.0824  962.0   
1  45.727478  ...         41.903530           30.615618  16.7081  988.0   
2  32.711426  ...         55.005234           50.175697   7.4865  942.0   
3  40.638546  ...         40.415058           34.756508   8.7289  970.0   
4  18.429998  ...         34.144085           35.040554   2.3091  950.0   

   sagree-effective  sagree-important  sagree-safe  sdisagree-effective  \
0          0.729730          0.937630     0.766112             0.008316   
1          0.715587          0.834008     0.609312             0.017206   
2          0.449045          0.585987     0.363057             0.031847   
3          0.501546          0.668557     0.393814             0.030928   
4          0.574737          0.840000     0.614737             0.012632   

   sdisagree-important  sdisagree-safe  
0             0.006237        0.013514  
1             0.018219        0.033401  
2             0.054140        0.058917  
3             0.022680        0.065979  
4             0.002105        0.014737  

[5 rows x 23 columns]

In [143]:
final_dset.dtypes

LOCATION                object
BUPPSRY-level          float64
TRY-level              float64
TRY_MEN-level          float64
TRY_WOMEN-level        float64
UPPSRY-level           float64
UPPSRY_MEN-level       float64
UPPSRY_WOMEN-level     float64
BUPPSRY-spend          float64
TRY-spend              float64
TRY_MEN-spend          float64
TRY_WOMEN-spend        float64
UPPSRY-spend           float64
UPPSRY_MEN-spend       float64
UPPSRY_WOMEN-spend     float64
C02                    float64
count                  float64
sagree-effective       float64
sagree-important       float64
sagree-safe            float64
sdisagree-effective    float64
sdisagree-important    float64
sdisagree-safe         float64
dtype: object